numpy/linalg/eig.cc

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "eig.h"
+#include <cstring>
+#include <iostream>
+
+Eig::Eig() {
+    a_mat = r_mat = vl_mat = vr_mat = wr_vec = wi_vec = 0;
+    vr_mat_complex = 0;
+}
+
+void Eig::make_args(int size) {
+    n = lda = ldvl = ldvr = size;
+
+    mat_size = n * n;
+
+    // input matrix
+    a_mat = make_random_mat(mat_size);
+    r_mat = make_mat(mat_size);
+
+    // left and right eigenvectors
+    vl_mat = make_mat(mat_size);
+    vr_mat = make_mat(mat_size);
+
+    // real and imaginary parts of eigenvalues
+    wr_vec = make_mat(n);
+    wi_vec = make_mat(n);
+
+    // complex eigenvectors
+    vr_mat_complex =
+        (double _Complex *) mkl_malloc(mat_size * sizeof(*vr_mat_complex), 64);
+}
+
+void Eig::copy_args() {
+    memcpy(r_mat, a_mat, mat_size * sizeof(*r_mat));
+    memset(vr_mat_complex, 0, mat_size * sizeof(*vr_mat_complex));
+}
+
+void Eig::compute() {
+    // Get right eigenvectors and eigenvalues
+    int info = LAPACKE_dgeev(LAPACK_COL_MAJOR, 'N', 'V', n, a_mat, lda, wr_vec,
+                             wi_vec, vl_mat, 1, vr_mat, ldvr);
+    assert(info == 0);
+
+    bool only_real = true;
+    for (int i = 0; i < n; i++) {
+        if (wi_vec[i] != 0.0) {
+            only_real = false;
+            break;
+        }
+    }
+
+    if (!only_real) {
+        // scipy.linalg.decomp._make_complex_eigvecs
+        for (int i = 0; i < n; i++) {
+            if (wi_vec[i] != 0.0) {
+                // Copy real and imaginary parts
+                for (int j = 0; j < n; j++) {
+                    vr_mat_complex[i * n + j] =
+                        CMPLX(vr_mat[i * n + j], vr_mat[(i + 1) * n + j]);
```

Comment:
is this part of scipy code?

Reply:
Yes, scipy has to transform LAPACK's output to something more user friendly. From what I understand, if there is a complex conjugate pair of eigenvectors, LAPACK will only return one, but will consume space for two eigenvectors, putting the real part of the `i`th eigenvector in `v[i]` and its imaginary part in `v[i+1]`. So scipy allocates new space for N complex eigenvectors and copies eigenvectors from LAPACK format to what the user expects.

- [Call to `_make_complex_eigvecs`]https://github.com/scipy/scipy/blob/v1.2.0/scipy/linalg/decomp.py#L260)
- [Definition of `_make_complex_eigvecs`](https://github.com/scipy/scipy/blob/v1.2.0/scipy/linalg/decomp.py#L36-L47)

Reply:
I'm not sure this code gets vectorized and it is certainly not parallel - thus I'm not sure how this can be an efficiency reference. Please add vectorization pragmas at least. Take a look out eig implementation in numpy

Reply:
The DGEEV call takes orders of magnitude more time than this extra computation, so I haven't been able to make a significant difference by adding vectorization pragmas or any other form of parallelism outside the DGEEV call.